### PR TITLE
add Emoji extension

### PR DESCRIPTION
**Pull request by FezVrasta**
## Description

A new extension to quickly find and copy emojis to clipboard.

Additional ideas I'd like to implement later on:

- [x] Recently used emojis
- [ ] Aliases (grinning face / face / smile / happy / joy / :D / grin)
- [ ] Skin tones selector

## Type of change

- New extension

## Screenshot

CleanShot 2021-10-14 at 14 17 46@2x

## Checklist

- [x] I read the [contribution guidelines](contributing.md)

### CODE REVIEW

**Comment by BProg** [comment]
It would be cool to add a press enter command to insert the emoji in the currently "focused" application's text input

File: extensions/emoji/package.json [hunk 1]
```diff
@@ -0,0 +1,36 @@
+{
+  "name": "emoji",
+  "title": "Emoji",
```
Comment (mattisssa) [inline extensions/emoji/package.json:3]: We recommend to use **Emoji Search** as the extension name

File: extensions/emoji/package.json [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  "description": "Finds an emoji and copy it to the clipboard",
+  "icon": "command-icon.png",
+  "author": "FezVrasta",
+  "license": "MIT",
+  "commands": [
+    {
+      "name": "emoji",
+      "title": "emoji",
```
Comment (mattisssa) [inline extensions/emoji/package.json:11]: And for the command name, it should be **Search Emoji**

File: extensions/emoji/src/emoji.tsx [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+      if (typeof cache === "string") {
+        if (!didUnmount.current) {
+          setState(JSON.parse(cache));
+        }
+      }
+    })();
+
+    return () => {
+      didUnmount.current = true;
```
Comment (mattisssa) [inline extensions/emoji/src/emoji.tsx:37]: Using `didUnmount` looks really complex, what are you using this for? To implement some kind of caching?

File: extensions/emoji/src/emoji.tsx [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+      list.find((x) => x.description === emoji.description) ? list : [emoji, ...list].slice(0, 10)
+    );
+  }, []);
+
+  return (
+    <List onSearchTextChange={setSearchText} isLoading={list.length === 0}>
+      {[{ category: "Recently Used", emojis: recentlyUsed }, ...list].map((category: Category) => (
+        <List.Section title={category.category} key={category.category}>
+          {category.emojis
+            .filter((emoji) => emoji.description.includes(searchText))
```
Comment (mattisssa) [inline extensions/emoji/src/emoji.tsx:81]: As far as I can see you are using the custom search which is completely similar to what we are doing in the built-in one. 
I'd recommend removing `onSearchTextChange` a filtering logic completely. It makes the searching works not smoothly.

And by the way, in the case, you need to provide special characters for the search, check out the `keywords` property for [List.Item](https://developers.raycast.com/api-reference/user-interface/list#props-1)

File: extensions/emoji/src/emoji.tsx (hunk 3 above)
Comment (FezVrasta) [reply extensions/emoji/src/emoji.tsx:37]: I just want to make sure the async code doesn't result in state being set on an unmounted component.

**Comment by FezVrasta** [comment]
@mattisssa I asked on Slack for a template to use, so that I get the shape, shadows, and size correct. They said it should be ready soon, I'll wait for it to prepare the icon, is that okay?

**Comment by PitNikola** [comment]
@FezVrasta, do you think you could something with jumping of "Recently Used" section? Maybe load everything in one go?

https://user-images.githubusercontent.com/4332171/140338276-c1a5d507-94d9-4bdb-8758-92dae0093806.mp4

**Comment by FezVrasta** [comment]
Yes I think I can wait for both of them to be loaded before displaying the list. I'm a bit busy lately but I'll send a PR next week.

**Pull request merged**